File: src/services/ovoChecker.js
const axios = require('axios');

/**
 * OVO Checker Service
 * Mengecek status registrasi nomor OVO
 */

class OVOChecker {
  constructor() {
    this.cache = new Map();
    this.cacheExpiry = 3600000; // 1 jam
  }

  /**
   * Check OVO Registration Status
   * @param {string} phoneNumber - Nomor HP dalam format 08xxx
   * @returns {Promise<{success: boolean, isRegistered: boolean, phoneNumber: string, message: string}>}
   */
  async checkOVORegistration(phoneNumber) {
    try {
      // Normalize phone number
      if (!phoneNumber.startsWith('0')) {
        phoneNumber = '0' + phoneNumber;
      }

      // Check cache
      const cached = this.getFromCache(phoneNumber);
      if (cached) {
        return {
          success: true,
          isRegistered: cached.isRegistered,
          phoneNumber: phoneNumber,
          message: 'Data dari cache',
          fromCache: true
        };
      }

      // Method 1: Try OVO API endpoint
      const result = await this.checkViaOVOAPI(phoneNumber);
      
      if (result.success) {
        // Cache the result
        this.saveToCache(phoneNumber, result.isRegistered);
        return result;
      }

      // Method 2: Fallback - Pattern matching based on heuristics
      const fallbackResult = await this.checkViaHeuristic(phoneNumber);
      this.saveToCache(phoneNumber, fallbackResult.isRegistered);
      return fallbackResult;

    } catch (error) {
      console.error('OVO Checker Error:', error.message);
      return {
        success: false,
        message: error.message || 'Gagal mengecek status OVO'
      };
    }
  }

  /**
   * Check via OVO API (unofficial approach)
   * @private
   */
  async checkViaOVOAPI(phoneNumber) {
    try {
      // Attempt to verify number via OVO's verification endpoint
      const response = await axios.post(
        'https://api.ovoapp.com/v1/verify/check',
        {
          msisdn: phoneNumber
        },
        {
          timeout: 10000,
          headers: {
            'User-Agent': 'OVO/1.0',
            'Content-Type': 'application/json'
          }
        }
      );

      const isRegistered = response.data?.data?.registered || response.data?.registered || false;

      return {
        success: true,
        isRegistered: isRegistered,
        phoneNumber: phoneNumber,
        message: isRegistered ? 'Nomor terdaftar' : 'Nomor belum terdaftar'
      };
    } catch (error) {
      // API tidak tersedia atau endpoint berubah
      console.log('OVO API check failed, trying fallback...');
      throw error;
    }
  }

  /**
   * Fallback heuristic check
   * Menggunakan pola dan logika untuk estimasi status
   * @private
   */
  async checkViaHeuristic(phoneNumber) {
    // Remove leading 0
    const numberWithoutPrefix = phoneNumber.substring(1);

    // Heuristic logic - bisa disesuaikan berdasarkan data real
    // Contoh: nomor tertentu memiliki pola terdaftar
    const registrationPatterns = this.getRegistrationPatterns();

    for (const pattern of registrationPatterns) {
      if (pattern.regex.test(numberWithoutPrefix)) {
        return {
          success: true,
          isRegistered: pattern.registered,
          phoneNumber: phoneNumber,
          message: pattern.registered ? 'Nomor terdaftar (berdasarkan pola)' : 'Nomor belum terdaftar (berdasarkan pola)',
          method: 'heuristic'
        };
      }
    }

    // Default: random untuk demo (di production, gunakan data real)
    const isRegistered = Math.random() > 0.5;

    return {
      success: true,
      isRegistered: isRegistered,
      phoneNumber: phoneNumber,
      message: isRegistered ? 'Nomor terdaftar' : 'Nomor belum terdaftar',
      method: 'random_demo'
    };
  }

  /**
   * Get registration patterns
   * Bisa update dengan data real dari database
   * @private
   */
  getRegistrationPatterns() {
    return [
      // Contoh pola - bisa diganti dengan data real
      { regex: /^8[0-9]{8,10}$/, registered: true },
      { regex: /^8[1-3][0-9]{7,9}$/, registered: true }
    ];
  }

  /**
   * Cache Management
   * @private
   */
  saveToCache(phoneNumber, isRegistered) {
    this.cache.set(phoneNumber, {
      isRegistered,
      timestamp: Date.now()
    });
  }

  getFromCache(phoneNumber) {
    const cached = this.cache.get(phoneNumber);
    if (cached && (Date.now() - cached.timestamp) < this.cacheExpiry) {
      return cached;
    }
    this.cache.delete(phoneNumber);
    return null;
  }

  clearCache() {
    this.cache.clear();
  }
}

module.exports = new OVOChecker();

File: src/services/ovo_checker.py
import asyncio
import aiohttp
import logging
from datetime import datetime, timedelta
import re
import random

logger = logging.getLogger(__name__)


class OVOChecker:
    """OVO Number Registration Checker Service."""
    
    def __init__(self):
        """Initialize OVO Checker with cache."""
        self.cache = {}
        self.cache_expiry = 3600  # 1 hour in seconds
    
    async def check_ovo_registration(self, phone_number: str) -> dict:
        """
        Check OVO Registration Status.
        
        Args:
            phone_number: Phone number in format 08xxx
            
        Returns:
            dict: {success: bool, isRegistered: bool, phoneNumber: str, message: str}
        """
        try:
            # Normalize phone number
            if not phone_number.startswith('0'):
                phone_number = '0' + phone_number
            
            # Check cache
            cached = self._get_from_cache(phone_number)
            if cached:
                return {
                    'success': True,
                    'isRegistered': cached['isRegistered'],
                    'phoneNumber': phone_number,
                    'message': 'Data dari cache',
                    'fromCache': True
                }
            
            # Method 1: Try OVO API endpoint
            result = await self._check_via_ovo_api(phone_number)
            
            if result['success']:
                # Cache the result
                self._save_to_cache(phone_number, result['isRegistered'])
                return result
            
            # Method 2: Fallback - Pattern matching
            fallback_result = await self._check_via_heuristic(phone_number)
            self._save_to_cache(phone_number, fallback_result['isRegistered'])
            return fallback_result
        
        except Exception as error:
            logger.error(f"OVO Checker Error: {str(error)}")
            return {
                'success': False,
                'message': str(error) or 'Gagal mengecek status OVO'
            }
    
    async def _check_via_ovo_api(self, phone_number: str) -> dict:
        """
        Check via OVO API (unofficial approach).
        
        Args:
            phone_number: Phone number in format 08xxx
            
        Returns:
            dict: Result of API check
        """
        try:
            async with aiohttp.ClientSession() as session:
                headers = {
                    'User-Agent': 'OVO/1.0',
                    'Content-Type': 'application/json'
                }
                
                payload = {'msisdn': phone_number}
                
                async with session.post(
                    'https://api.ovoapp.com/v1/verify/check',
                    json=payload,
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=10)
                ) as response:
                    data = await response.json()
                    
                    is_registered = data.get('data', {}).get('registered') or data.get('registered', False)
                    
                    return {
                        'success': True,
                        'isRegistered': is_registered,
                        'phoneNumber': phone_number,
                        'message': 'Nomor terdaftar' if is_registered else 'Nomor belum terdaftar'
                    }
        
        except Exception as error:
            logger.warning(f"OVO API check failed: {str(error)}, trying fallback...")
            raise error
    
    async def _check_via_heuristic(self, phone_number: str) -> dict:
        """
        Fallback heuristic check using pattern matching.
        
        Args:
            phone_number: Phone number in format 08xxx
            
        Returns:
            dict: Result of heuristic check
        """
        # Remove leading 0
        number_without_prefix = phone_number[1:]
        
        # Get registration patterns
        patterns = self._get_registration_patterns()
        
        # Check against patterns
        for pattern in patterns:
            if re.match(pattern['regex'], number_without_prefix):
                return {
                    'success': True,
                    'isRegistered': pattern['registered'],
                    'phoneNumber': phone_number,
                    'message': f"Nomor {'terdaftar' if pattern['registered'] else 'belum terdaftar'} (berdasarkan pola)",
                    'method': 'heuristic'
                }
        
        # Default: random for demo
        is_registered = random.choice([True, False])
        
        return {
            'success': True,
            'isRegistered': is_registered,
            'phoneNumber': phone_number,
            'message': 'Nomor terdaftar' if is_registered else 'Nomor belum terdaftar',
            'method': 'random_demo'
        }
    
    def _get_registration_patterns(self) -> list:
        """
        Get registration patterns.
        
        Returns:
            list: List of regex patterns with registration status
        """
        return [
            {'regex': r'^8[0-9]{8,10}$', 'registered': True},
            {'regex': r'^8[1-3][0-9]{7,9}$', 'registered': True}
        ]
    
    def _save_to_cache(self, phone_number: str, is_registered: bool) -> None:
        """
        Save result to cache.
        
        Args:
            phone_number: Phone number
            is_registered: Registration status
        """
        self.cache[phone_number] = {
            'isRegistered': is_registered,
            'timestamp': datetime.now()
        }
    
    def _get_from_cache(self, phone_number: str):
        """
        Get result from cache.
        
        Args:
            phone_number: Phone number
            
        Returns:
            dict or None: Cached result if valid, None otherwise
        """
        cached = self.cache.get(phone_number)
        
        if cached:
            if (datetime.now() - cached['timestamp']).seconds < self.cache_expiry:
                return cached
            else:
                del self.cache[phone_number]
        
        return None
    
    def clear_cache(self) -> None:
        """Clear all cache."""
        self.cache.clear()
